### PR TITLE
perf scripts updates

### PR DESCRIPTION
**Pull request by malay-nagda**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Updated perf configs for various llms

**Collection**: [llm]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: scripts/performance/llm/pretrain_llama31_405b.py [hunk 1]
```diff
@@ -89,6 +89,16 @@ def override_recipe_configs(
     # data module configs
     recipe.data.tokenizer = hf_tokenizer("meta-llama/Llama-3.1-405B")
 
+    userbuffers_bf16_h100_h16384_tp8_cp2_mbs1_seqlen8192.qkv_fprop.aggregate = False
```
Comment (guyueh1) [inline scripts/performance/llm/pretrain_llama31_405b.py:92]: this can be removed since we've changed the userbuffers.py?

Comment (malay-nagda) [reply scripts/performance/llm/pretrain_llama31_405b.py:92]: For now, I think we should keep in both places.
The change in userbuffers.py will not be available in 25.04, but users can still use this from latest perf scripts.

We can remove this from next NeMo container when usebuffers.py chnage is picked up?

Comment (guyueh1) [reply scripts/performance/llm/pretrain_llama31_405b.py:92]: I think as long as the userbuffers.py is changed and people use the latest NeMo repo to launch jobs, then they will get the updated config. Because at launch time all the details of tp overlap are saved in the generated config.

Comment (malay-nagda) [reply scripts/performance/llm/pretrain_llama31_405b.py:92]: Ah yes! I'll update it

Comment (malay-nagda) [reply scripts/performance/llm/pretrain_llama31_405b.py:92]: More nuance:
I can update this. But, we have to ask users to do step 2.-
1. NeMo "package" installed in the venv is used for resolving these args.
2. User has to install latest NeMo "package" in the venv to use the change in userbuffers.py. Simply using latest NeMo to launch the job will not work.
3. But, if the change is in perf script, the override will still work even if the NeMo package in venv is older (I usually pin to the same commit as inside container to ensure we are not setting/unsetting configs unintentionally)

Comment (guyueh1) [reply scripts/performance/llm/pretrain_llama31_405b.py:92]: Yes, user need to do 2
Then I think we can keep these

**Comment by guyueh1** [comment]
@malay-nagda can you add the`PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` to the `custom_env_vars` for pretrain deepseek v3? It turns out to be the key factor between OOM and functional for the B200 EP8 setting.

**Comment by malay-nagda** [comment]
> @malay-nagda can you add the`PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` to the `custom_env_vars` for pretrain deepseek v3? It turns out to be the key factor between OOM and functional for the B200 EP8 setting.

That is already added in PerfEnvPlugin @guyueh1

**Comment by malay-nagda** [comment]
@ko3n1g @chtruong814 can you help merge this PR?

**Comment by chtruong814** [comment]
Test failures are likely from the transformers package update we encountered earlier and unrelated to these changes.  Will merge this now.
